mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1666,6 +1666,72 @@ struct TosaFoldConstantPad : public TosaFoldConstantBase<tosa::PadOp> {
   }
 };
 
+template <typename BaseType>
+DenseElementsAttr tile(ShapedType inputType, ElementsAttr inputValues,
+                       ShapedType outputType, BaseType zero) {
+  auto inputShape = inputValues.getShapedType().getShape();
+  auto outputShape = outputType.getShape();
+
+  SmallVector<BaseType> outputValues(outputType.getNumElements(), zero);
+  SmallVector<int64_t> indexInTarget(outputType.getRank());
+
+  auto values = inputValues.getValues<BaseType>();
+  for (size_t outIndex = 0, e = outputValues.size(); outIndex < e; ++outIndex) {
+    auto index = offsetToIndex(outputShape, outIndex);
+    for (auto i = 0; i < outputType.getRank(); ++i) {
+      indexInTarget[i] = index[i] % inputShape[i];
+    }
+    auto inputIndexOffset = indexToOffset(inputShape, indexInTarget);
+    BaseType value = values[inputIndexOffset];
+    outputValues[outIndex] = value;
+  }
+  return DenseElementsAttr::get(outputType,
+                                llvm::ArrayRef<BaseType>(outputValues));
+}
+
+struct TosaFoldConstantTile : public TosaFoldConstantBase<tosa::TileOp> {
+  using TosaFoldConstantBase::TosaFoldConstantBase;
+
+  LogicalResult matchAndRewrite(tosa::TileOp op,
+                                PatternRewriter &rewriter) const override {
+    auto outputType = cast<ShapedType>(op.getType());
+    // TOSA doesn't support quantized types.
+    if (!outputType.getElementType().isIntOrIndexOrFloat())
+      return failure();
+
+    auto input = op.getInput1();
+    ElementsAttr inputValues;
+    if (!matchPattern(input, m_Constant(&inputValues)))
+      return failure();
+
+    auto inputType = inputValues.getShapedType();
+    // Only fold op with multiple users if foldSplatOrSingleUseOnly == true.
+    if (!llvm::hasSingleElement(input.getDefiningOp()->getUsers()) &&
+        foldSplatOrSingleUseOnly)
+      return failure();
+
+    auto elementType = inputValues.getElementType();
+
+    DenseElementsAttr resultAttr;
+    // Handle integer types with APInt
+    if (auto intType = dyn_cast<IntegerType>(elementType))
+      resultAttr = tile<APInt>(inputType, inputValues, outputType,
```

Comment:
Converting to APInt/APFloat and back will dominate this folding in runtime because the actual folding is just copying.
Can you look at the optimizations that are done in the transpose folding where the actual types `float`, ... are used?